tests/functional/csv_generator.py

```diff
@@ -0,0 +1,123 @@
+"""
+A Hypothesis strategy for generating valid CSV files
+"""
+
+import functools
+import string
+from csv import writer
+from io import StringIO
+
+from hypothesis.strategies import (
+    composite,
+    floats,
+    integers,
+    lists,
+    sampled_from,
+    text,
+)
+
+
+def _records_to_csv(rows, lineterminator):
+    """
+    Convert the results into a csv string
+    """
+    f = StringIO()
+    w = writer(f, lineterminator=lineterminator)
+    for row in rows:
+        w.writerow(row)
+
+    return f.getvalue()
+
+
+def _escape_csv_field(field) -> str:
+    field = str(field)
+
+    # Carriage return characters (\r) become newline (\n) characters in the html
+    # This is fine, but annoying for the tests which then fail. So easier to just
+    # exclude them as a possible csv character
+    while "\r" in field:
+        field = field.replace("\r", "_")
```

Comment:
I think the loop is unnecessary:

```
>>> "a\rb\rc".replace("\r", "_")
'a_b_c'
```

Reply:
I was thinking that the field might contain `\\rr`, in which case the first replace would still leave `\r`.

Reply:
ah - but that was when I was replacing it with the empty string. Replacing with the underscore almost certainly means it's redundant.